### PR TITLE
[CALCITE-3282] HiveSqlDialect unparse Interger type as Int in order to be compatible with Hive1.x

### PR DESCRIPTION
**Pull request by ffmax**
For example, a sql "select cast('1' as int)", HiveSqlDialect unparse it to "select cast('1' as Integer)".  However, "Integer" is not  allowed in  hive1.x,  for consideration of compatibility, it should be unparse as "int". Related jira is https://issues.apache.org/jira/browse/CALCITE-3282.

### CODE REVIEW

**Comment by julianhyde** [comment]
Can you add the JIRA case number?

**Comment by julianhyde** [comment]
There's some refactoring to be done. There is already a `SqlDialect.getCastSpec` method. That method needs to be documented (and if necessary refactored). 

It's crazy that your fix works by overriding `SqlDialect.unparseCall` when there is already a `SqlDialect.getCastSpec`, and there a lot of other code in `SqlImplementor.toSql(RexProgram program, RexNode rex)`.

As part of your fix, can you sort out this mess, and make it a bit better rather than a bit worse?

**Comment by ffmax** [comment]
Sorry, JIRA case number has been added,  and actually,  I ignore SqlDialect.getCastSpec method and I will try fix it better.

**Comment by danny0405** [comment]
@julianhyde @ffmax I want to clarify that it is me that suggests to modify the `SqlDialect#unparseCall` method, previously i didn't pay attention to the `SqlDialect.getCastSpec` method which i think is the right place that need modification. I think we should add some detailed documentation for method `SqlDialect.getCastSpec`.

Sorry for that @ffmax !

**Comment by danny0405** [comment]
Another question: it seems that Hive 2.x already supports `Integer`, but we can not distinguish the Hive version here.

**Comment by ffmax** [comment]
I find It might be necessary to add  INT enum in  SqlTypeName, if we want to refactor getCastSpec to do with the issue.

**Comment by ffmax** [comment]
@danny0405 Yeah, Hive 2.x already supports Integer, but using the common type between Hive1.x and Hive2.x may be better?

File: core/src/main/java/org/apache/calcite/sql/dialect/HiveSqlDialect.java [hunk 1]
```diff
@@ -149,6 +127,22 @@ private void unparseTrim(SqlWriter writer, SqlCall call, int leftPrec,
...
   }
+
+  @Override public SqlNode getCastSpec(final RelDataType type) {
+    if (type instanceof BasicSqlType) {
+      switch (type.getSqlTypeName()) {
+      case INTEGER:
+        final SqlBasicTypeNameSpec typeNameSpec = new SqlBasicTypeNameSpec(
+            INT,
+            type.getPrecision(),
+            type.getScale(),
```
Comment (danny0405) [inline core/src/main/java/org/apache/calcite/sql/dialect/HiveSqlDialect.java]: Use `SqlUserDefinedTypeNameSpec` and pass in the name you want to use instead, there is no need to add a new `SqlTypeName`.

File: core/src/main/java/org/apache/calcite/sql/SqlDialect.java [hunk 2]
```diff
@@ -749,6 +749,9 @@ public boolean supportsDataType(RelDataType type) {
     return true;
   }
 
+ /** Returns SqlNode for type in "cast(column as type)" , which is different
```
Comment (jinxing64) [inline core/src/main/java/org/apache/calcite/sql/SqlDialect.java]: Not sure, shall we wrap `cast ...` with `{@code}`?

Comment (jinxing64) [reply core/src/main/java/org/apache/calcite/sql/SqlDialect.java]: > which is different in every kind of database such as type name, precision etc.

How about:
which might be different between databases by type name, precision etc.

File: core/src/main/java/org/apache/calcite/sql/type/SqlTypeUtil.java [hunk 3]
```diff
@@ -1009,6 +1009,7 @@ private static boolean flattenFields(
     return nested;
   }
 
+
   /**
```
Comment (jinxing64) [inline core/src/main/java/org/apache/calcite/sql/type/SqlTypeUtil.java]: No need to add this line ?

File: core/src/main/java/org/apache/calcite/sql/type/SqlTypeUtil.java [hunk 4]
```diff
@@ -1054,7 +1055,7 @@ public static SqlDataTypeSpec convertTypeToSpec(RelDataType type,
    * @param type type descriptor
    * @return corresponding parse representation
    */
-  public static SqlDataTypeSpec convertTypeToSpec(RelDataType type) {
+  public static SqlDataTypeSpec  convertTypeToSpec(RelDataType type) {
```
Comment (jinxing64) [inline core/src/main/java/org/apache/calcite/sql/type/SqlTypeUtil.java]: No necessary change ?

File: core/src/main/java/org/apache/calcite/sql/dialect/HiveSqlDialect.java (hunk 1 above)
Comment (ffmax) [reply core/src/main/java/org/apache/calcite/sql/dialect/HiveSqlDialect.java]: Thx

File: core/src/main/java/org/apache/calcite/sql/SqlDialect.java (hunk 2 above)
Comment (ffmax) [reply core/src/main/java/org/apache/calcite/sql/SqlDialect.java]: > Not sure, shall we wrap `cast ...` with `{@code}`?

Actually, I think it is not easily understand either.

File: core/src/main/java/org/apache/calcite/sql/type/SqlTypeUtil.java (hunk 3 above)
Comment (ffmax) [reply core/src/main/java/org/apache/calcite/sql/type/SqlTypeUtil.java]: Sorry,  I made some modification and recover it , but left new line and space.

**Comment by jinxing64** [comment]
Not sure, is it necessary to squash the commits when merge ?

**Pull request closed without merging**